### ISSUE TITLE
Support for other compositors (KDE, Gnome, ...)

### ISSUE THREAD

**Issue by Ferdi265** [Ferdi265/wl-mirror#17]
There are a few things that need to be done for wl-mirror to work on other compositors than sway / wlroots:

- [ ] Implement XDG Desktop Portal backend (#16)
- [ ] Add support for requesting server side decorations with XDG Decorations or the KDE Decoration Protocol
- [ ] Add client-side decorations for Gnome / Mutter (which does not support XDG Decorations)

The last of these is likely going to be a long way off, if it is ever implemented. Since wl-mirror does not use a toolkit, adding client-side decorations would involve a whole new level of complexity since that would require handling mouse input and seats.

**Comment by Ferdi265** [comment]
For client-side decorations, [libdecor](https://gitlab.freedesktop.org/libdecor/libdecor) could also be an option, if it's easily possible to add it to our event loop unintrusively.

**Comment by Ferdi265** [comment]
I started implementing libdecor support for server and client side window decorations. Current progress tracked in branch [feature-libdecor](https://github.com/Ferdi265/wl-mirror/tree/feature-libdecor)

**Comment by Ferdi265** [comment]
It seems I found the fix for the flickering resize on sway: I was committing the libdecor frame and resizing the EGL surface in the wrong order.

I think this now just needs a small bit of cleanup, and then decoration support is done, ticking two of the above boxes.

**Comment by Ferdi265** [comment]
With libdecor and the xdg portals backend, wl-mirror will now work on KDE and Gnome!

Both branches are still experimental and aren't finalized yet, but apparently this started working just on its own!

## KDE

![kde-select](https://github.com/user-attachments/assets/603d2fa8-7591-4044-8afe-edb6ea1f5882)
![kde-mirror](https://github.com/user-attachments/assets/207fc151-1b1d-47bd-bce3-6a7cd7771739)

## Gnome

![gnome-select](https://github.com/user-attachments/assets/5399a96b-e91c-404d-b33e-c07cdd9a96b0)
![gnome-mirror](https://github.com/user-attachments/assets/3ce192a0-ee5d-4213-89a8-f64a6a7fbdeb)

**Comment by Ferdi265** [comment]
libdecor support is now merged into main.

**Comment by nerdictranslator** [comment]
So, quick heads up, KDE can now mirror displays while preserving the aspect ratio of the original and the frame rates of both. Tested on KDE 6.6, mirroring my superultrawide monitor to the capture card in my streaming PC.

I'd still love to have a way to do this for single windows, and there are advantages to using a separate window for the mirrored display, but it's cool to see this feature implemented in KDE natively.

<img width="3840" height="2160" alt="Image" src="https://github.com/user-attachments/assets/f21cfc11-715d-4607-a3e3-6e3dd2495c00" />

**Comment by Ferdi265** [comment]
> So, quick heads up, KDE can now mirror displays while preserving the aspect ratio of the original and the frame rates of both. Tested on KDE 6.6, mirroring my superultrawide monitor to the capture card in my streaming PC.
> 
> I'd still love to have a way to do this for single windows, and there are advantages to using a separate window for the mirrored display, but it's cool to see this feature implemented in KDE natively.
> <img alt="Image" width="2000" height="2000" src="https://private-user-images.githubusercontent.com/208233699/556242725-f21cfc11-715d-4607-a3e3-6e3dd2495c00.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzIyOTUxNjcsIm5iZiI6MTc3MjI5NDg2NywicGF0aCI6Ii8yMDgyMzM2OTkvNTU2MjQyNzI1LWYyMWNmYzExLTcxNWQtNDYwNy1hM2UzLTZlM2RkMjQ5NWMwMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIyOFQxNjA3NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNWZhZWRkYzg4Mzg4YmMyZGY3ODlkOWExM2ZhMzA2NjQwM2U2MTk0NTBlYTBhNzAyYjVmZjU0MWM2NTE5ZWNiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BTpcLlm06o2Fdcj8HgW6vYdVd5vwbNH0Jc-kgUuz9tg">

Great! I still plan on porting wl-mirror, and once ported it will also support mirroring individual windows with various scaling settings, but that might still be a while until it's done, since I don't have much time for side projects like this one at the moment.

**Comment by nerdictranslator** [comment]
@Ferdi265 Fair enough, and thank you for all you've done already. It's a really cool project.
